File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.cpp
class Solution {
public:
    int solve(int i,int k,vector<int> &arr ,vector<int> &dp){
        if(i>=arr.size()) return 0;
        int &ret=dp[i];
        if(~ret) return ret;
        ret=-2e9;
        int mx=-2e9;
        for(int j=i;j<min((int)arr.size(),i+k);j++){
            mx=max(mx,arr[j]);
            ret=max(ret,mx*(j-i+1)+solve(j+1,k,arr,dp));
        }
        return ret;
    }
    int maxSumAfterPartitioning(vector<int>& arr, int k) {
        vector<int> dp(arr.size()+1,-1);
        return solve(0,k,arr,dp);
        
    }
};

File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.java
class Solution {

    int Solve(int a[],int l , int k,int n,int dp[] ){
        if(l>=n) return 0 ; 
        int ret = dp[l];
        if(ret!=-1) return ret ; 
        int mx = a[l];
        for(int i = l ;i<Math.min(n,l+k);i++){
            mx = Math.max(a[i],mx);
            ret = Math.max(ret , mx*(i-l+1)+Solve(a,i+1,k,n,dp));
        }

        return dp[l] = ret ; 
    }

    public int maxSumAfterPartitioning(int[] arr, int k) {
        int n = arr.length;
        int dp[] = new int[n];
        Arrays.fill(dp,-1);
        return Solve(arr ,0,k,n , dp);
    }
}